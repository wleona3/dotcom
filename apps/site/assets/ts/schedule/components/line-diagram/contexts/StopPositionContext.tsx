import React, {
  createContext,
  MutableRefObject,
  PropsWithChildren,
  useCallback,
  useContext,
  useLayoutEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { LineDiagramStop } from "../../__schedule";
import {
  BASE_LINE_WIDTH,
  BRANCH_SPACING,
  StopCoord
} from "../graphics/graphic-helpers";
import { isMergeStop } from "../line-diagram-helpers";

type refType = HTMLElement | null;
export interface StopCoordState {
  [k: string]: StopCoord | null;
}
interface ContextValues {
  stopCoordState: StopCoordState;
  resetAllCoordinates: () => void;
  stopRefMap: Map<string, refType>;
}

/**
 * Computes pixel coordinates for a line diagram stop bubble
 * May be null if the stop is hidden on a collapsed branch
 */
function calcCoordinates(
  el: HTMLElement,
  stop: LineDiagramStop
): StopCoord | null {
  // x usually just a function of stop_data length
  const x = isMergeStop(stop)
    ? BASE_LINE_WIDTH + 1
    : BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
  // y should align with the ref position
  const { offsetTop, offsetHeight } = el;
  const y = offsetTop + offsetHeight / 2;
  return x && y ? [x, y] : null;
}

/**
 * Sets up a ref for every stop,
 * sets up state to hold coordinate values,
 * sets up a function to update coordinates based on the refs.
 */
const useStopPositions = (stops: LineDiagramStop[]): ContextValues => {
  const map = new Map(
    stops.map<[string, refType]>(stop => [stop.route_stop.id, null])
  );
  const objEntries = stops.map<[string, StopCoord | null]>(stop => [
    stop.route_stop.id,
    null
  ]);
  const stopRefMap: MutableRefObject<Map<string, refType>> = useRef(map);
  const initialObject = Object.fromEntries(objEntries);
  const [stopCoordState, setStopCoordState] = useState<StopCoordState>(
    initialObject
  );

  const resetAllCoordinates = useCallback((): void => {
    window.requestAnimationFrame((): void => {
      // get new coordinates
      setStopCoordState(
        (state): StopCoordState => {
          const newState = { ...state };
          stops.forEach(stop => {
            const stopId = stop.route_stop.id;
            const el = stopRefMap.current.get(stopId);
            newState[stopId] = el ? calcCoordinates(el, stop) : null;
          });
          return newState;
        }
      );
    });
  }, [stops, stopRefMap]);

  // only update when the state actually changes (not every render)
  const contextValues = useMemo(
    () => ({
      stopCoordState,
      resetAllCoordinates,
      stopRefMap: stopRefMap.current
    }),
    [stopCoordState, resetAllCoordinates]
  );

  // Automatically update positions after resize events
  useLayoutEffect(() => {
    window.addEventListener("resize", resetAllCoordinates);
    return () => window.removeEventListener("resize", resetAllCoordinates);
  }, [resetAllCoordinates]);

  return contextValues;
};

/**
 * Sets up context that will hold:
 * stopCoordState: an object containing updated values for pixel location
 * resetAllCoordinates: a function to recalculate coordinates for every stop
 */
export const StopPositionContext = createContext<ContextValues>({
  stopCoordState: {},
  resetAllCoordinates: () => {},
  stopRefMap: new Map()
});

// Get latest coordinates for one or all stops
export function useStopPositionCoordinates(): StopCoordState | null;
export function useStopPositionCoordinates(stopId?: string): StopCoord;
export function useStopPositionCoordinates(
  stopId?: string
): StopCoordState | StopCoord | null {
  const { stopCoordState } = useContext(StopPositionContext);
  if (stopId) {
    return stopCoordState[stopId];
  }
  return stopCoordState;
}

// Triggers recalculation of all coordinates
export function useStopPositionReset(): () => void {
  const { resetAllCoordinates } = useContext(StopPositionContext);
  return resetAllCoordinates;
}

// Returns the set of refs being used to track <h4> stop names
export const useStopRefs = (): Map<string, refType> => {
  const { stopRefMap } = useContext(StopPositionContext);
  return stopRefMap;
};

/**
 * Access context values within child components via
 *  useStopPositionCoordinates(stopId?)
 *  useStopPositionReset()
 *  useStopPositionRefs()
 */
export const StopPositionProvider = ({
  stops,
  children
}: PropsWithChildren<{ stops: LineDiagramStop[] }>): React.ReactElement => {
  const contextValues = useStopPositions(stops);
  return (
    <StopPositionContext.Provider value={contextValues}>
      {children}
    </StopPositionContext.Provider>
  );
};
