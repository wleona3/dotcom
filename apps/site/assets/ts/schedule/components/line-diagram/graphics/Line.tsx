import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { LineDiagramStop } from "../../__schedule";
import {
  areOnDifferentBranchLines,
  isMergeStop,
  isBranchTerminusStop
} from "../line-diagram-helpers";
import {
  StopCoord,
  CoordState,
  BRANCH_LINE_WIDTH,
  BASE_LINE_WIDTH,
  BRANCH_SPACING
} from "./graphic-helpers";

interface PathGraphicsProps {
  from: LineDiagramStop;
  to: LineDiagramStop;
}

const showCRLine = (from: LineDiagramStop, to: LineDiagramStop): boolean => {
  const OLNoCR = ["place-haecl", "place-state","place-dwnxg","place-chncl","place-tumnl"]
  return !(OLNoCR.includes(from.route_stop.id) || OLNoCR.includes(to.route_stop.id));
}

const showShuttleLine = (from: LineDiagramStop, to: LineDiagramStop): boolean => {
  const OLNoShuttle = ["place-state","place-dwnxg","place-chncl","place-tumnl"]
  return !(OLNoShuttle.includes(from.route_stop.id) || OLNoShuttle.includes(to.route_stop.id));
}
const Line = ({ from, to }: PathGraphicsProps): ReactElement | null => {
  const fromCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[from.route_stop.id]
  );
  const toCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[to.route_stop.id]
  );
  if (!fromCoords || !toCoords) return null;
  if (isMergeStop(to) && from.stop_data.length === to.stop_data.length) {
    return null;
  }

  const [x1, y1] = fromCoords;
  const y2 = toCoords[1];

  if (isMergeStop(from) && !areOnDifferentBranchLines(from, to)) {
    // just draw one line.
    const lineProps: React.SVGProps<SVGLineElement> = {
      key: `${from.route_stop.id}-${to.route_stop.id}`
    };
    lineProps.strokeWidth = `${BASE_LINE_WIDTH}px`;
    lineProps.x1 = `${x1}px`;
    lineProps.x2 = lineProps.x1;
    lineProps.y1 = `${y1}px`;
    lineProps.y2 = `${y2}px`;
    // if (
    //   // from.stop_data.some(sd => sd["has_disruption?"]) &&
    //   // to.stop_data.some(sd => sd["has_disruption?"])
    //   true
    // ) {
    //   lineProps.stroke = "url(#diagonalHatch)";
    // }
    lineProps.stroke = "url(#diagonalHatch)";
    return <line className="line-diagram-svg__line" {...lineProps} />;
  }

  // otherwise let's just show all the lines...
  return (
    <>
      {from.stop_data.map((branchLine, branchIndex) => {
        const lineProps: React.SVGProps<SVGLineElement> = {
          key: `${from.route_stop.id}-${to.route_stop.id}-${branchLine.type}-${branchIndex}`
        };
        lineProps.stroke = "url(#diagonalHatch)";
        lineProps.strokeWidth = `${
          branchIndex >= 1 &&
          !(isBranchTerminusStop(from) || isBranchTerminusStop(to)) &&
          (branchLine.type === "stop" ||
            to.stop_data[branchIndex].type === "stop")
            ? BRANCH_LINE_WIDTH
            : BASE_LINE_WIDTH
        }px`;
        lineProps.x1 = `${BRANCH_SPACING * branchIndex +
          BASE_LINE_WIDTH +
          1}px`;
        lineProps.x2 = lineProps.x1;
        lineProps.y1 = `${y1}px`;
        lineProps.y2 = `${y2}px`;
        // if (branchLine["has_disruption?"]) {
        //   lineProps.stroke = "url(#diagonalHatch)";
        // }

        const linePropsCR = { ...lineProps };
        linePropsCR.x1 = `${BRANCH_SPACING * (branchIndex + 1) +
          BASE_LINE_WIDTH +
          1}px`;
        linePropsCR.x2 = linePropsCR.x1;
        linePropsCR.stroke = "purple";
        return <>
          {showShuttleLine(from, to) ? <line key={lineProps.key} className="line-diagram-svg__line" {...lineProps} /> : null}
          {showCRLine(from, to) ? <line key={`${from.route_stop.id}-${to.route_stop.id}-${branchLine.type}-${branchIndex + 1}--cr`} className="line-diagram-svg__line u-text--commuter-rail" {...linePropsCR} />: null}
        </>;
      })}
    </>
  );
};

export default Line;
