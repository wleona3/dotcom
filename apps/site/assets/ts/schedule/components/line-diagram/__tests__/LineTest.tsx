import React from "react";
import * as redux from "react-redux";
import { cloneDeep, last } from "lodash";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import {
  BASE_LINE_WIDTH,
  BRANCH_SPACING,
  BRANCH_LINE_WIDTH
} from "../graphics/graphic-helpers";
import Line from "../graphics/Line";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";
import { StopCoordState } from "../contexts/StopPositionContext";
const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
const lineDiagramWithBranching = (outwardLineDiagram as unknown) as LineDiagramStop[];

const [from, to] = lineDiagram.slice(0, 2);
const [fromWithBranching, toWithBranching] = lineDiagramWithBranching.slice(
  4,
  6
);
const [testX, testY] = [7, 17];

// mock the redux state
const mockState: StopCoordState = {
  [from.route_stop.id]: [testX, testY],
  [to.route_stop.id]: [testX, testY + 7],
  [fromWithBranching.route_stop.id]: [testX, testY],
  [toWithBranching.route_stop.id]: [testX, testY + 7]
};

describe("Line component", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeAll(() => {
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <svg>
        <Line from={from} to={to} />
      </svg>,
      [from, to],
      mockState
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(
      container.querySelector("line.line-diagram-svg__line")
    ).toBeDefined();
  });

  it("standard line has expected props", () => {
    const lineElement = container.querySelector("line.line-diagram-svg__line")!;
    expect(lineElement.getAttribute("strokeWidth")).toEqual(
      `${BASE_LINE_WIDTH}px`
    );
    expect(lineElement.getAttribute("x1")).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(lineElement.getAttribute("x2")).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(lineElement.getAttribute("y1")).toEqual(`${testY}px`);
    expect(lineElement.getAttribute("y2")).toEqual(`${testY + 7}px`);
  });
});

describe("Line component between stops with branches", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeAll(() => {
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <svg>
        <Line from={fromWithBranching} to={toWithBranching} />
      </svg>,
      [fromWithBranching, toWithBranching],
      mockState
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(container.querySelector("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("line on branch has expected props", () => {
    const lineElement = container.querySelector("line.line-diagram-svg__line")!;
    expect(lineElement.getAttribute("strokeWidth")).toEqual(
      `${BRANCH_LINE_WIDTH}px`
    );
    expect(lineElement.getAttribute("x1")).toEqual(
      `${BRANCH_SPACING * 2 + BASE_LINE_WIDTH + 1}px`
    );
    expect(lineElement.getAttribute("x2")).toEqual(
      `${BRANCH_SPACING * 2 + BASE_LINE_WIDTH + 1}px`
    );
    expect(lineElement.getAttribute("y1")).toEqual(`${testY}px`);
    expect(lineElement.getAttribute("y2")).toEqual(`${testY + 7}px`);
  });
});

describe("Line component between stops with disruptions", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeAll(() => {
    const fromWithDetour = cloneDeep(from);
    last(fromWithDetour.stop_data)!["has_disruption?"] = true;
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <svg>
        <Line from={fromWithDetour} to={to} />
      </svg>,
      [fromWithDetour, to],
      mockState
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows an SVG line between the stops", () => {
    expect(container.querySelector("line.line-diagram-svg__line")).toBeTruthy();
  });

  it("has expected props, including stroke pattern", () => {
    const lineElement = container.querySelector("line.line-diagram-svg__line")!;
    expect(lineElement.getAttribute("strokeWidth")).toEqual(
      `${BASE_LINE_WIDTH}px`
    );
    expect(lineElement.getAttribute("x1")).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(lineElement.getAttribute("x2")).toEqual(`${BASE_LINE_WIDTH + 1}px`);
    expect(lineElement.getAttribute("y1")).toEqual(`${testY}px`);
    expect(lineElement.getAttribute("y2")).toEqual(`${testY + 7}px`);
    expect(lineElement.getAttribute("stroke")).toContain("url(#shuttle)");
  });
});
