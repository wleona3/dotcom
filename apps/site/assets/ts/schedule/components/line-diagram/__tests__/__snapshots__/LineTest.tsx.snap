// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line component between stops with branches renders and matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <line
      class="line-diagram-svg__line"
      stroke-width="10px"
      x1="11px"
      x2="11px"
      y1="17px"
      y2="24px"
    />
    <line
      class="line-diagram-svg__line"
      stroke-width="10px"
      x1="27px"
      x2="27px"
      y1="17px"
      y2="24px"
    />
    <line
      class="line-diagram-svg__line"
      stroke-width="16px"
      x1="43px"
      x2="43px"
      y1="17px"
      y2="24px"
    />
  </svg>
</DocumentFragment>
`;

exports[`Line component between stops with disruptions renders and matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <line
      class="line-diagram-svg__line"
      stroke="url(#shuttle)"
      stroke-width="10px"
      x1="11px"
      x2="11px"
      y1="17px"
      y2="24px"
    />
  </svg>
</DocumentFragment>
`;

exports[`Line component renders and matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <line
      class="line-diagram-svg__line"
      stroke="url(#shuttle)"
      stroke-width="10px"
      x1="11px"
      x2="11px"
      y1="17px"
      y2="24px"
    />
  </svg>
</DocumentFragment>
`;
